### PR TITLE
switch from volta to fnm

### LINKED ISSUE #47441: Revert use of Volta for Dockerfile
Volta doesn't work on ARM: "Error: Sorry! Volta currently only provides pre-built binaries for x86_64 architectures". We can't have that be part of the default Dockerfile, then. Images must be usable on ARM without modifications.

So have to revert https://github.com/rails/rails/pull/46794.

### PR DESCRIPTION
**Pull request by rubys**
Fixes #47441

See also https://github.com/rails/rails/pull/47446#issuecomment-1438599148

### CODE REVIEW

**Comment by zzak** [comment]
Tried this against my branch from #47453, which allows me to vendor rails from a checkout into Docker.

Perhaps it is me, but I ran into this issue when running this command:

```
$ bundle exec railties/exe/rails new ~/code/railsdev --javascript=rollup --vendor
$ cd ~/code/railsdev && docker build .
```

Results in the following error:

```
 => ERROR [build 10/10] RUN SECRET_KEY_BASE_DUMMY=1 ./bin/rails assets:precompile            1.1s
------
 > [build 10/10] RUN SECRET_KEY_BASE_DUMMY=1 ./bin/rails assets:precompile:
#23 0.931 yarn install v1.22.19
#23 0.951 [1/4] Resolving packages...
#23 0.962 success Already up-to-date.
#23 0.966 Done in 0.04s.
#23 1.056 yarn run v1.22.19
#23 1.076 error Command "build" not found.
#23 1.076 info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
#23 1.085 rails aborted!
#23 1.085 jsbundling-rails: Command build failed, ensure yarn is installed and `yarn build` runs without errors
#23 1.085 /rails/vendor/rails/railties/lib/rails/commands/rake/rake_command.rb:28:in `block (2 levels) in perform'
#23 1.085 /rails/vendor/rails/railties/lib/rails/commands/rake/rake_command.rb:28:in `block in perform'
#23 1.085 /rails/vendor/rails/railties/lib/rails/commands/rake/rake_command.rb:18:in `perform'
#23 1.085 /rails/vendor/rails/railties/lib/rails/command.rb:152:in `invoke_rake'
#23 1.085 /rails/vendor/rails/railties/lib/rails/command.rb:73:in `block in invoke'
#23 1.085 /rails/vendor/rails/railties/lib/rails/command.rb:145:in `with_argv'
#23 1.085 /rails/vendor/rails/railties/lib/rails/command.rb:69:in `invoke'
#23 1.085 /rails/vendor/rails/railties/lib/rails/commands.rb:18:in `<main>'
#23 1.085 Tasks: TOP => assets:precompile => javascript:build
#23 1.085 (See full trace by running task with --trace)
------
executor failed running [/bin/sh -c SECRET_KEY_BASE_DUMMY=1 ./bin/rails assets:precompile]: exit code: 1
```

I'd love to know what I'm doing wrong, or how to properly test this stuff in a checkout. :bow:

**Comment by rubys** [comment]
I'll check into it in the morning (I'm east coast US), but will share what I did to test this:

```
rails new welcome --main --javascript=esbuild
cd welcome
```

From there, I manually edited the rails entry in the Gemfile to look like:

```
gem "rails", github: "rails/rails", branch: "main"
```

From there, I tried both `docker build .` and also for good measure deployed to fly.io.

This change also matches what `dockerfile-rails` produces.  And while all of fly.io machines are amd64 machines, there are a number of demos that you can run yourself on a M1/M2 machine, including this one: https://github.com/rubys/dockerfile-rails/blob/main/DEMO.md#demo-5---bunding-javascript-esbuild - try it with rails 7.0.4.2.

**Comment by rubys** [comment]
OK, before I turn in for the night:

```
#23 1.076 error Command "build" not found.
#23 1.076 info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

This generally means that your `package.json` doesn't have a script named build.  Can you check?  I see nothing in either my change or your vendor change that would cause this message.

**Comment by zzak** [comment]
I've tried the method you mentioned above, but I'll note that `--main` will build an app against main, which still has volta in the Dockerfile. Only running with `--dev` would copy the files as they are here, when checked out.

It does make sense that `build` task is missing from `package.json`, in fact it is pretty bare:

```json
{
  "name": "app",
  "private": "true",
  "dependencies": {
    "@hotwired/stimulus": "^3.2.1",
    "@hotwired/turbo-rails": "^7.2.5",
    "esbuild": "^0.17.10"
  }
}
```

Interesting is that this appears in the output from running the `rails new` command:

```
Add build script
Add "scripts": { "build": "esbuild app/javascript/*.* --bundle --sourcemap --outdir=app/assets/builds --public-path=assets" } to your package.json
       rails  turbo:install stimulus:install
```

I have no idea where that comes from however. Sorry if I've lead us astray here. 🤔

**Comment by rubys** [comment]
>  but I'll note that --main will build an app against main,

Sorry, late night thinko :-).  I did verify that the generated Dockerfile was as expected before deploying.

A vendor option seems useful.

File: railties/lib/rails/generators/rails/app/templates/Dockerfile.tt [hunk 1]
```diff
@@ -23,9 +23,10 @@ RUN apt-get update -qq && \
 # Install JavaScript dependencies
 ARG NODE_VERSION=<%= dockerfile_node_version %>
 ARG YARN_VERSION=<%= dockerfile_yarn_version %>
-ENV VOLTA_HOME="/usr/local"
-RUN curl https://get.volta.sh | bash && \
-    volta install node@$NODE_VERSION yarn@$YARN_VERSION
+RUN curl -fsSL https://fnm.vercel.app/install | bash && \
+    /root/.local/share/fnm/fnm install $NODE_VERSION
```
Comment (byroot) [inline railties/lib/rails/generators/rails/app/templates/Dockerfile.tt:27]: I wonder if we really need a version manager for this. I looked at the implementation and it basically just download a tarball from https://nodejs.org/dist/ and extract it.

I don't really see a reason not to do that ourselves and skip a middleman.

**Pull request closed without merging**